tiledb/api/c_api/profile/profile_api_experimental.h

```diff
@@ -0,0 +1,156 @@
+ * Users may specifiy this path, or use `nullptr` for the default.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_profile_t* profile;
+ * tiledb_error_t* error = NULL;
+ * tiledb_profile_alloc(
+ *   "my_profile",
+ *   TemporaryLocalDirectory("unit_capi_profile").path(),
+ *   &profile,
+ *   &error);
+ *
+ * tiledb_profile_t* profile1;
+ * tiledb_error_t* error1 = NULL;
+ * tiledb_profile_alloc(nullptr, nullptr, &profile1, &error1);
+ * @endcode
+ *
+ * @param[in] name The profile name, or `nullptr` for default.
+ * @param[in] homedir The path to `$HOME` directory, or `nullptr` for default.
+ * @param[out] profile The profile object to be created.
+ * @param[out] error Error object returned upon error (`NULL` if there is
+ *     no error).
+ * @return `TILEDB_OK` for success and `TILEDB_OOM` or `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT capi_return_t tiledb_profile_alloc(
+    const char* name,
+    const char* homedir,
+    tiledb_profile_t** profile,
+    tiledb_error_t** error) TILEDB_NOEXCEPT;
+
+/**
+ * Frees a TileDB profile object.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_profile_t* profile;
+ * tiledb_error_t* error = NULL;
+ * tiledb_profile_alloc(&profile, &error);
+ * tiledb_profile_set_param("rest.username", "my_username");
+ * tiledb_profile_save();
+ * tiledb_profile_free(&profile);
+ * @endcode
+ *
+ * @param[in] profile The profile object to be freed.
+ */
+TILEDB_EXPORT void tiledb_profile_free(tiledb_profile_t** profile)
+    TILEDB_NOEXCEPT;
+
+/**
+ * Retrieves the name from the given profile.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_profile_t* profile;
+ * tiledb_error_t* error = NULL;
+ * tiledb_profile_alloc(&profile, &error);
+ * tiledb_string_t* name;
+ * tiledb_profile_get_name(profile, &name);
+ * tiledb_profile_free(&profile);
+ * @endcode
+ *
+ * @param[in] profile The profile.
+ * @param[out] name The name of the profile, to be retrieved.
+ * @param[out] error Error object returned upon error (`NULL` if there is
+ *     no error).
+ * @return TILEDB_EXPORT
+ */
+TILEDB_EXPORT capi_return_t tiledb_profile_get_name(
+    tiledb_profile_t* profile,
+    tiledb_string_t** name,
+    tiledb_error_t** error) TILEDB_NOEXCEPT;
+
+/**
+ * Retrieves the homedir from the given profile.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_profile_t* profile;
+ * tiledb_error_t* error = NULL;
+ * tiledb_profile_alloc(&profile, &error);
+ * tiledb_string_t* homedir;
+ * tiledb_profile_get_homedir(profile, &homedir);
+ * tiledb_profile_free(&profile);
+ * @endcode
+ *
+ * @param[in] profile The profile.
+ * @param[out] homedir The homedir of the profile, to be retrieved.
+ * @param[out] error Error object returned upon error (`NULL` if there is
+ *     no error).
+ * @return TILEDB_EXPORT
+ */
+TILEDB_EXPORT capi_return_t tiledb_profile_get_homedir(
+    tiledb_profile_t* profile,
+    tiledb_string_t** homedir,
+    tiledb_error_t** error) TILEDB_NOEXCEPT;
+
```

Comment:
Do we need a getter for the home directory? In the overwhelming majority of production use cases, people would not set a custom home directory and just use the default one. In fact we added custom home directory support for tests.

Reply:
I feel rather strongly that the "profile directory" *should* be an argument here.

We may encounter a situation where N users try to share the same profile.  By hard-coding a `$HOME` directory we are preventing that use case.  By allowing a "profile directory" argument we are enabling users who want this set up to manage it themselves on a shared file system.

It's a great coincidence that we have a similar need for testing.  Even if the need described above is not a real one we already built the functionality and the burden for users to understand what passing `nullptr` as an argument does is not a high one.

Reply:
Additionally, this is a consequence of a choice to consolidate what was previously separate "user" and "test" APIs.

Reply:
> We may encounter a situation where N users try to share the same profile […] on a shared file system

My understanding is that this is not an expected scenario. Profiles contain sensitive access credentials that are not supposed to be shared by multiple users.